### ISSUE TITLE
Spack can't fetch libgd, url handling seems to be problematic

### ISSUE THREAD

**Issue by hartzell** [LLNL/spack#3554]
libgd won't build, it fails trying to fetch the library (for all three versions in the package.py).

```
[hartzelg@lb093login spack]$ spack install libgd
==> Installing libgd
==> libtiff is already installed in /home/hartzelg/tmp/spack/opt/spack/linux-centos7-x86_64/gcc-5.4.0/libtiff-4.0.6-g7gpsqww4hfkg5tg4nr5izmc2sj76c2z
==> [...]
==> pkg-config is already installed in /home/hartzelg/tmp/spack/opt/spack/linux-centos7-x86_64/gcc-5.4.0/pkg-config-0.29.1-ld3i7xv7jtacp7nt7k3aimjleiiuhpdg
==> Fetching https://github.com/libgd/libgd/releases/download/2.2.4/lib2.2.4.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from https://github.com/libgd/libgd/releases/download/2.2.4/lib2.2.4.tar.gz failed.
```

Here's the URL from the package.py:

```python
    url      = 'https://github.com/libgd/libgd/releases/download/gd-2.2.4/libgd-2.2.4.tar.gz'
```

Note that the url in `package.py` contains this bit: 

```
.../download/gd-2.2.4/libgd...
```
but the URL that's being fetched looks like this: 

```
.../download/2.2.4/lib2...
```

In particular, the one being fetched is lacking the `gd-` in two places.

Ring any bells?

**Comment by hartzell** [comment]
Ok, I think that the problem is that the package is named `libgd` but there's a thing that looks like a version number that has *just* `gd-` in front of it.  That's taken as part of the version in the demo url and is excised when the url is built.

I can fix it in this case by defining a `url_for_version` for libgd.  Is there a wider issue here or should I just PR the libgd change?

**Comment by adamjstewart** [comment]
Just submit a PR with a url_for_version. #2972 will probably fix this. Keep the issue open and I'll double check.

**Comment by adamjstewart** [comment]
Can confirm that this issue will be solved by #2972.

### Before
```
$ spack url parse https://github.com/libgd/libgd/releases/download/gd-2.2.4/libgd-2.2.4.tar.gz
==> Parsing URL: https://github.com/libgd/libgd/releases/download/gd-2.2.4/libgd-2.2.4.tar.gz

==> Matched version regex  7: r'github.com/[^/]+/[^/]+/releases/download/v?([^/]+)/.*$'
==> Matched  name   regex  8: r'([^/]+)gd-2.2.4'

==> Detected:
    https://github.com/libgd/libgd/releases/download/gd-2.2.4/libgd-2.2.4.tar.gz
                                                     ~~~~~~~~ ---
    name:    lib
    version: gd-2.2.4

==> Substituting version 9.9.9b:
    https://github.com/libgd/libgd/releases/download/9.9.9b/lib9.9.9b.tar.gz
                                                     ~~~~~~ ---
```
### After
```
$ spack url parse https://github.com/libgd/libgd/releases/download/gd-2.2.4/libgd-2.2.4.tar.gz
==> Parsing URL: https://github.com/libgd/libgd/releases/download/gd-2.2.4/libgd-2.2.4.tar.gz

==> Matched version regex  0: r'^[a-zA-Z+._-]+[._-]v?(\\d[\\d._-]*)$'
==> Matched  name   regex  0: r'github\\.com/[^/]+/([^/]+)'

==> Detected:
    https://github.com/libgd/libgd/releases/download/gd-2.2.4/libgd-2.2.4.tar.gz
                             -----                                  ~~~~~
    name:    libgd
    version: 2.2.4

==> Substituting version 9.9.9b:
    https://github.com/libgd/libgd/releases/download/gd-9.9.9b/libgd-9.9.9b.tar.gz
                             -----                                   ~~~~~~
```
